
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Tables documentation &#8212; The 214K ARG metagenomic data collection</title>
    
  <link href="_static/css/theme.css" rel="stylesheet">
  <link href="_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Data loading" href="00_Data_loading.html" />
    <link rel="prev" title="Welcome" href="intro.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">The 214K ARG metagenomic data collection</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="intro.html">
   Welcome
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Tables documentation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="00_Data_loading.html">
   Data loading
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/Tables.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/hmmartiny/mARG"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/hmmartiny/mARG/issues/new?title=Issue%20on%20page%20%2FTables.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#metadata">
   Metadata
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#arg-counts">
   ARG counts
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#resfinder-annotations">
     ResFinder annotations
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#rrna-counts">
   rRNA counts
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#diversity-measures">
   Diversity measures
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Tables documentation</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#metadata">
   Metadata
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#arg-counts">
   ARG counts
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#resfinder-annotations">
     ResFinder annotations
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#rrna-counts">
   rRNA counts
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#diversity-measures">
   Diversity measures
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="tables-documentation">
<h1>Tables documentation<a class="headerlink" href="#tables-documentation" title="Permalink to this headline">¶</a></h1>
<p>The data available at [ZENODO LINK] consists of five different tables, which will be introduced on this page. This data collection consists of <a class="reference internal" href="#metadata"><span class="std std-ref">metadata</span></a> for each of the sequencing datasets we retrieved, as well as the results of aligning reads to reference sequences from the <a class="reference internal" href="#counts-arg"><span class="std std-ref">ResFinder</span></a> and <a class="reference internal" href="#counts-silva"><span class="std std-ref">Silva</span></a> databases and accompanying <a class="reference internal" href="#diversity"><span class="std std-ref">diversity measurements</span></a>.</p>
<div class="section" id="metadata">
<span id="id1"></span><h2>Metadata<a class="headerlink" href="#metadata" title="Permalink to this headline">¶</a></h2>
<table class="table">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Column</p></th>
<th class="head"><p>Explanation</p></th>
<th class="head"><p>Example</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>run_accession</p></td>
<td><p>Identifer for sequencing reads.</p></td>
<td><p>DRR000836</p></td>
</tr>
<tr class="row-odd"><td><p>sample_accession</p></td>
<td><p>Identifer for sample.</p></td>
<td><p>SAMD00002573</p></td>
</tr>
<tr class="row-even"><td><p>project_accession</p></td>
<td><p>Identifer for project.</p></td>
<td><p>PRJDA61421</p></td>
</tr>
<tr class="row-odd"><td><p>country</p></td>
<td><p>Locality of sample isolation: country names, oceans or seas, followed by regions and localities.</p></td>
<td><p>N/A</p></td>
</tr>
<tr class="row-even"><td><p>location</p></td>
<td><p>Geographic location of isolation of the sample (latitude, longitude).</p></td>
<td><p>N/A</p></td>
</tr>
<tr class="row-odd"><td><p>continent</p></td>
<td><p>Geographic region (land or water body)</p></td>
<td><p>N/A</p></td>
</tr>
<tr class="row-even"><td><p>collection_date</p></td>
<td><p>Date that the specimen was collected.</p></td>
<td><p>2016-01-01</p></td>
</tr>
<tr class="row-odd"><td><p>tax_id</p></td>
<td><p>NCBI taxon ID of the organism from which the sample was obtained.</p></td>
<td><p>939928</p></td>
</tr>
<tr class="row-even"><td><p>host</p></td>
<td><p>Natural (as opposed to laboratory) host to the organism from which sample was obtained.</p></td>
<td><p>Homo sapiens</p></td>
</tr>
<tr class="row-odd"><td><p>host_tax_id</p></td>
<td><p>NCBI taxon ID of the host</p></td>
<td><p>9606</p></td>
</tr>
<tr class="row-even"><td><p>instrument_platform</p></td>
<td><p>Instrument platform used in sequencing experiment.</p></td>
<td><p>LS454</p></td>
</tr>
<tr class="row-odd"><td><p>instrument_model</p></td>
<td><p>Instrument model used in sequencing experiment.</p></td>
<td><p>454 GS FLX Titanium</p></td>
</tr>
<tr class="row-even"><td><p>library_layout</p></td>
<td><p>Sequencing library layout.</p></td>
<td><p>SINGLE</p></td>
</tr>
<tr class="row-odd"><td><p>raw_reads</p></td>
<td><p>Number of raw sequencing reads.</p></td>
<td><p>1268608</p></td>
</tr>
<tr class="row-even"><td><p>trimmed_reads</p></td>
<td><p>Number of trimmed sequencing reads.</p></td>
<td><p>1247751</p></td>
</tr>
<tr class="row-odd"><td><p>raw_bases</p></td>
<td><p>Number of bases in the raw sequencing reads.</p></td>
<td><p>641025182</p></td>
</tr>
<tr class="row-even"><td><p>trimmed_bases</p></td>
<td><p>Number of bases in the trimmed sequencing reads.</p></td>
<td><p>1247751</p></td>
</tr>
<tr class="row-odd"><td><p>trimmed_fragments</p></td>
<td><p>Number of trimmed read fragments that can be mapped.</p></td>
<td><p>1247751</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="arg-counts">
<span id="counts-arg"></span><h2>ARG counts<a class="headerlink" href="#arg-counts" title="Permalink to this headline">¶</a></h2>
<table class="table">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Column</p></th>
<th class="head"><p>Explanation</p></th>
<th class="head"><p>Example</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>run_accession</p></td>
<td><p>Identifer for sequencing reads.</p></td>
<td><p>DRR000836</p></td>
</tr>
<tr class="row-odd"><td><p>sample_accession</p></td>
<td><p>Identifer for sample.</p></td>
<td><p>SAMD00002573</p></td>
</tr>
<tr class="row-even"><td><p>project_accession</p></td>
<td><p>Identifer for project.</p></td>
<td><p>PRJDA61421</p></td>
</tr>
<tr class="row-odd"><td><p>trimmed_fragments</p></td>
<td><p>Total number of fragments.</p></td>
<td><p>1247751</p></td>
</tr>
<tr class="row-even"><td><p>run_date</p></td>
<td><p>Date that KMA program ran.</p></td>
<td><p>2020-11-10</p></td>
</tr>
<tr class="row-odd"><td><p>kma_version</p></td>
<td><p>Version of KMA used.</p></td>
<td><p>1.3.0</p></td>
</tr>
<tr class="row-even"><td><p>refSequence</p></td>
<td><p>Name of template sequence.</p></td>
<td><p>blaACT-4_2_AJ311172</p></td>
</tr>
<tr class="row-odd"><td><p>refSequence_length</p></td>
<td><p>Length of template sequence in bp.</p></td>
<td><p>1146</p></td>
</tr>
<tr class="row-even"><td><p>refCoveredPositions</p></td>
<td><p>The number of covered positions in the template with a minimum depth of 1.</p></td>
<td><p>427</p></td>
</tr>
<tr class="row-odd"><td><p>refConsensusSum</p></td>
<td><p>Total number of bases identical to the template.</p></td>
<td><p>424</p></td>
</tr>
<tr class="row-even"><td><p>bpTotal</p></td>
<td><p>Total number of bases aligned to the template.</p></td>
<td><p>1651</p></td>
</tr>
<tr class="row-odd"><td><p>fragmentCountAln</p></td>
<td><p>Number of fragments mapped and aligned to the template.</p></td>
<td><p>5</p></td>
</tr>
<tr class="row-even"><td><p>bacterial_fragment</p></td>
<td><p>Sum of rRNA fragments mappped and aligned for the dataset.</p></td>
<td><p>1428</p></td>
</tr>
</tbody>
</table>
<div class="section" id="resfinder-annotations">
<span id="arg-anno"></span><h3>ResFinder annotations<a class="headerlink" href="#resfinder-annotations" title="Permalink to this headline">¶</a></h3>
<p>To group ARGs by classes, phenotypes or mechanisms, one can use the annotation table below. It contains the annotation given in the official documentation for the ResFinder database (<a class="reference external" href="https://bitbucket.org/genomicepidemiology/resfinder_db">link</a>).</p>
<table class="table">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Column</p></th>
<th class="head"><p>Explanation</p></th>
<th class="head"><p>Example</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>gene</p></td>
<td><p>Name of gene (refSequence)</p></td>
<td><p>aac(2’)-Ia_1_L06156</p></td>
</tr>
<tr class="row-odd"><td><p>anno_type</p></td>
<td><p>What type of label is given for the gene (Class, gene_length, Phenotype, Mechanism)</p></td>
<td><p>Class</p></td>
</tr>
<tr class="row-even"><td><p>anno_value</p></td>
<td><p>The label for the annotation type</p></td>
<td><p>Aminoglycoside</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="rrna-counts">
<span id="counts-silva"></span><h2>rRNA counts<a class="headerlink" href="#rrna-counts" title="Permalink to this headline">¶</a></h2>
<table class="table">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Column</p></th>
<th class="head"><p>Explanation</p></th>
<th class="head"><p>Example</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>run_accession</p></td>
<td><p>Identifer for sequencing reads.</p></td>
<td><p>DRR000836</p></td>
</tr>
<tr class="row-odd"><td><p>sample_accession</p></td>
<td><p>Identifer for sample.</p></td>
<td><p>SAMD00002573</p></td>
</tr>
<tr class="row-even"><td><p>project_accession</p></td>
<td><p>Identifer for project.</p></td>
<td><p>PRJDA61421</p></td>
</tr>
<tr class="row-odd"><td><p>run_date</p></td>
<td><p>Date that KMA program ran.</p></td>
<td><p>2020-11-10</p></td>
</tr>
<tr class="row-even"><td><p>kma_version</p></td>
<td><p>Version of KMA used.</p></td>
<td><p>1.3.0</p></td>
</tr>
<tr class="row-odd"><td><p>phylum_name</p></td>
<td><p>Name of phylum.</p></td>
<td><p>Bacteroidetes</p></td>
</tr>
<tr class="row-even"><td><p>phylum_tax</p></td>
<td><p>NCBI taxon ID of the phylum.</p></td>
<td><p>976</p></td>
</tr>
<tr class="row-odd"><td><p>genus_name</p></td>
<td><p>Name of genus.</p></td>
<td><p>Myroides</p></td>
</tr>
<tr class="row-even"><td><p>genus_tax</p></td>
<td><p>NCBI taxon ID of the genus.</p></td>
<td><p>76831</p></td>
</tr>
<tr class="row-odd"><td><p>fragmentCountAln</p></td>
<td><p>Sum of fragments aligned to rRNA genes belonging to the corresponding genus.</p></td>
<td><p>10.7744</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="diversity-measures">
<span id="diversity"></span><h2>Diversity measures<a class="headerlink" href="#diversity-measures" title="Permalink to this headline">¶</a></h2>
<table class="table">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Column</p></th>
<th class="head"><p>Explanation</p></th>
<th class="head"><p>Example</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>run_accession</p></td>
<td><p>Identifer for sequencing reads.</p></td>
<td><p>DRR000836</p></td>
</tr>
<tr class="row-odd"><td><p>kma_version</p></td>
<td><p>Version of KMA used.</p></td>
<td><p>1.3.0</p></td>
</tr>
<tr class="row-even"><td><p>category</p></td>
<td><p>What group of genes the diversity measures are for: ARG, Phyla or Genera.</p></td>
<td><p>ARG</p></td>
</tr>
<tr class="row-odd"><td><p>total_fragments</p></td>
<td><p>Total fragments that could be mapped.</p></td>
<td><p>1247751</p></td>
</tr>
<tr class="row-even"><td><p>category_fragments</p></td>
<td><p>Number of fragments aligned to the category genes.</p></td>
<td><p>6.6925</p></td>
</tr>
<tr class="row-odd"><td><p>n</p></td>
<td><p>Number of unique genes/taxon IDs in group (observed richness)</p></td>
<td><p>3</p></td>
</tr>
<tr class="row-even"><td><p>Shannon</p></td>
<td><p>Shannon diversity index for category</p></td>
<td><p>0.887324</p></td>
</tr>
<tr class="row-odd"><td><p>Simpson</p></td>
<td><p>Simpson (1-D) diversity index for category</p></td>
<td><p>0.604705</p></td>
</tr>
</tbody>
</table>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="intro.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Welcome</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="00_Data_loading.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Data loading</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By Hannah-Marie Martiny<br/>
    
        &copy; Copyright 2022.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>